app/ota_client.py

```diff
@@ -873,177 +1062,136 @@ def _create_symbolic_links(self, list_file, standby_path):
 
-        if ishardlink and not hardlink_first_copy:
+        # if is_hardlink file, get a tracker from the register
+        is_hardlink = reginf.nlink >= 2
+        if is_hardlink:
+            _hardlink_tracker, _is_writer = hardlink_register.get_tracker(
+                reginf.sha256hash, reginf.path, reginf.nlink
+            )
+
+        # case 1: is hardlink and this thread is subscriber
+        if is_hardlink and not _is_writer:
             # wait until the first copy is ready
-            hardlink_event.wait()
-            (standby_path / prev_reginf.path.relative_to("/")).link_to(dst)
-            processed["op"] = "link"
-            processed["errors"] = 0
-        else:  # normal file or first copy of hardlink file
-            if reginf.path.is_file() and verify_file(
-                reginf.path, reginf.sha256hash, reginf.size
-            ):
-                # copy file from active bank if hash is the same
-                shutil.copy(reginf.path, dst)
-                processed["op"] = "copy"
-                processed["errors"] = 0
-            else:
-                processed["errors"] = downloader(
-                    url_base,
-                    reginf.path,
-                    dst,
-                    reginf.sha256hash,
-                    cookies=cookies,
-                )
-                processed["op"] = "download"
+            prev_reginf_path = _hardlink_tracker.subscribe()
+            (standby_path / prev_reginf_path.relative_to("/")).link_to(dst)
+            processed.op = "link"
+
+        # case 2: normal file or first copy of hardlink file
+        else:
+            try:
+                if reginf.path.is_file() and verify_file(
+                    reginf.path, reginf.sha256hash, reginf.size
+                ):
+                    # copy file from active bank if hash is the same
+                    shutil.copy(reginf.path, dst)
+                    processed.op = "copy"
+                else:
+                    processed.errors = downloader(
+                        reginf.path,
+                        dst,
+                        reginf.sha256hash,
+                    )
+                    processed.op = "download"
+
+                if is_hardlink and _is_writer:
+                    # writer indicates that the first copy of hardlink is ready
+                    _hardlink_tracker.writer_done()
+            except Exception:
+                # signal all subscribers to abort
+                if is_hardlink and _is_writer:
+                    _hardlink_tracker.writer_on_failed()
+
+                raise
 
-        processed["size"] = dst.stat().st_size
+        processed.size = dst.stat().st_size
 
+        # set file permission as RegInf says
         os.chown(dst, reginf.uid, reginf.gid)
         os.chmod(dst, reginf.mode)
 
-        end_time = time.time()
-        processed["elapsed"] = end_time - begin_time
+        processed.elapsed = time.thread_time() - begin_time
```

Comment:
In the current implementation, the actual time is returned. But this implementation, sleep time is reduced. The actual elapsed time should be returned.

Reply:
About this point, I think sleep time should not be counted.
When thread is sleeping, it doesn't mean the whole program is halted, only this thread becomes interrupt-able and releases GIL, so that other thread can take the place to execute. The execution time elapsed of different threads are overlapped.

In other word, during a thread's sleeping, other threads can also do jobs. So it is inaccurate to use real world time elapsed to measure this thread's time cost. Using real world time elapsed will result in duplicated counting of execution time from different threads. 

In summary, I think `total_elapsed_time` should use `time.time()` or `time.perf_count()` to calculate, actual execution time of each task(by thread worker) should use `time.thread_time()` to calculate.

Reply:
I will mention this in the source code as comment, thank you!